### PR TITLE
net: buf: Move net_buf_pool objects to dedicated linker area

### PR DESCRIPTION
**Pull request by jhedberg**
Moving the net_buf_pool objects to a dedicated area lets us access
them by array offset into this area instead of directly by pointer.
This helps reduce the size of net_buf objects by 4 bytes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

### CODE REVIEW

File: include/net/buf.h [hunk 1]
```diff
@@ -590,7 +601,9 @@ struct net_buf *net_buf_get(struct k_fifo *fifo, s32_t timeout);
  */
 static inline void net_buf_destroy(struct net_buf *buf)
 {
-	k_lifo_put(&buf->pool->free, buf);
+	struct net_buf_pool *pool = net_buf_pool_get(buf->pool_id);
```
Comment (jukkar) [inline include/net/buf.h]: If we are not checking the return value of net_buf_pool_get(), then why not write

k_lifo_put(&net_buf_pool_get(buf->pool_id)->free, buf)

instead and save extra variable here.

File: subsys/bluetooth/host/conn.c [hunk 2]
```diff
@@ -1055,9 +1055,12 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, u8_t flags)
 int bt_conn_send_cb(struct bt_conn *conn, struct net_buf *buf,
 		    bt_conn_tx_cb_t cb)
 {
+	struct net_buf_pool *pool;
+
 	BT_DBG("conn handle %u buf len %u cb %p", conn->handle, buf->len, cb);
 
-	if (buf->pool->user_data_size < BT_BUF_USER_DATA_MIN) {
+	pool = net_buf_pool_get(buf->pool_id);
+	if (pool->user_data_size < BT_BUF_USER_DATA_MIN) {
```
Comment (jukkar) [inline subsys/bluetooth/host/conn.c]: Same comment here, the pool variable looks useless.

**Review by jukkar — COMMENTED** [review]
Minor comment, I am also fine with what you propose.

File: include/net/buf.h (hunk 1 above)
Comment (jhedberg) [inline include/net/buf.h]: No other reason except code clarity. Presumably the compiler will generate the same code either way with the default optimization level (I'd need to verify that however).

**Review by jukkar — APPROVED** [review]
LGTM

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**